### PR TITLE
Fspca components lower bound

### LINKED ISSUE #588: Lower bound for FSPCA and bispectrum components
`fspca_components` and `bispectrum_components` in `RIRClass2D` could both use a lower bound check.

For extremely low resolution, like you might try for faster unit tests, you may have less than the current default of 400 fspca coefficients in your (FB) basis.  This will cause some errors later.  You could elect to warn and take some minimum (the number of FB basis coef), _or_ just raise at this point instead of having a more confusing error later and changing the unit tests to prescribe a smaller number of components.

### PR DESCRIPTION
**Pull request by garrettwrong**
Closes #588 

Also significantly speeds up Class2D unit tests by removing redundant tests and changing some configuration. (~20m to ~1m).

~Need to review coverage and potentially add some more sanity checks.~ (covered)

### CODE REVIEW

**Comment by garrettwrong** [comment]
@chris-langfield , could you take a look before I send this to Joakim? Thanks

File: src/aspire/basis/fspca.py [hunk 1]
```diff
@@ -31,15 +31,19 @@ class FSPCABasis(SteerableBasis2D):
...
-    def __init__(self, src, basis=None, noise_var=None, components=400, batch_size=512):
+    def __init__(
+        self, src, basis=None, noise_var=None, components=None, batch_size=512
+    ):
         """
 
         :param src: Source instance
         :param basis: Optional Fourier Bessel Basis (usually FFBBasis2D)
+        :param components: Number of components (compression).
+        Defaults None implies basis.count, therefor no compression.
```
Comment (chris-langfield) [inline src/aspire/basis/fspca.py]: Maybe reword this param docstring for clarity + "therefor"

File: src/aspire/basis/fspca.py [hunk 2]
```diff
@@ -31,15 +31,19 @@ class FSPCABasis(SteerableBasis2D):
...
         :param noise_var: None estimates noise (default).
-        :param batch_size: Batch size for computing basis coefficients.
         0 forces "clean" treatment (no weighting).
         Other values assigned to noise_var.
```
Comment (chris-langfield) [inline src/aspire/basis/fspca.py]: I know you didn't touch this line but maybe the blurb for `noise_var` could be reworded? Not sure what the second part is trying to say

File: src/aspire/classification/rir_class2d.py [hunk 3]
```diff
@@ -148,6 +113,54 @@ def __init__(
...
+        # For now, only run with FSPCA basis
+        if pca_basis and not isinstance(pca_basis, FSPCABasis):
+            raise NotImplementedError(
+                "RIRClass2D has currently only been developed for pca_basis as a FSPCABasis."
+            )
+        self.pca_basis = pca_basis
+
+        # When a user provides a basis, the fspca_components arg is either
+        # redudant (same) or conflicting (different).
+        # So when a user provides fspca_components, we need to check it matches the basis' compoenents.
```
Comment (chris-langfield) [inline src/aspire/classification/rir_class2d.py]: *components

File: src/aspire/classification/rir_class2d.py [hunk 4]
```diff
@@ -148,6 +113,54 @@ def __init__(
...
 
+        # For now, only run with FSPCA basis
+        if pca_basis and not isinstance(pca_basis, FSPCABasis):
+            raise NotImplementedError(
+                "RIRClass2D has currently only been developed for pca_basis as a FSPCABasis."
+            )
+        self.pca_basis = pca_basis
+
+        # When a user provides a basis, the fspca_components arg is either
+        # redudant (same) or conflicting (different).
```
Comment (chris-langfield) [inline src/aspire/classification/rir_class2d.py]: redundant

File: src/aspire/classification/rir_class2d.py [hunk 5]
```diff
@@ -148,6 +113,54 @@ def __init__(
...
+                )
+        elif pca_basis:  # fspca_components not specfied (default to taking from basis)
+            fspca_components = pca_basis.components
+
+            # For small problems, such as unit tests, we also need to guard against
+            #   requesting more fspca_components than exist in the basis now,
+            # In the case RIRClass2d instantiates the pca_basis,
+            #   this will be checked then, in FSPCABasis.
+            pca_basis._check_components()
+
```
Comment (chris-langfield) [inline src/aspire/classification/rir_class2d.py:147]: So the user can provide `pca_basis`, `fspca_components`, both, or neither, and the result would be:

`pca_basis` only: `fspca_components` is set to `pca_basis.count`
`fspca_components` only: `fspca_components` can be anything greater than `bispectrum_components`
Both: `fspca_components` must be equal to `pca_basis.components` but if the user inputs 400 it's set to `pca_basis.components` silently
Neither: `fspca_components = 400`

(Wrote it out mostly for myself to understand) Just wanted to confirm whether we still want the 400 default since we had gotten rid of it in `FSPCABasis`?

File: tests/test_class2D.py [hunk 6]
```diff
@@ -152,6 +180,7 @@ def testIncorrectComponents(self):
                 self.clean_src,
                 self.clean_fspca_basis,  # 400 components
                 fspca_components=100,
+                bispectrum_components=100,
                 large_pca_implementation="legacy",
```
Comment (chris-langfield) [inline tests/test_class2D.py]: Why do we have to add this? Shouldn't this not be needed since 400 > 300 and the bispectrum check comes later anyway?

File: tests/test_class2D.py [hunk 7]
```diff
@@ -276,7 +307,10 @@ def testImplementations(self):
         # Nearest Neighbhor component
         with pytest.raises(ValueError, match=r"Provided nn_implementation.*"):
             _ = RIRClass2D(
-                self.clean_src, self.clean_fspca_basis, nn_implementation="badinput"
+                self.clean_src,
+                self.clean_fspca_basis,
+                bispectrum_components=int(0.75 * self.clean_fspca_basis.basis.count),
```
Comment (chris-langfield) [inline tests/test_class2D.py:315]: Is this line just to avoid the `fspca_components < bispectrum_components` error?

Comment (garrettwrong) [reply tests/test_class2D.py:315]: Yeah, it's a hint at reasonable value instead of hard coding it. I can hard code it if you think that is cleaner.

File: src/aspire/basis/fspca.py (hunk 1 above)
Comment (garrettwrong) [reply src/aspire/basis/fspca.py]: Sure.

File: src/aspire/basis/fspca.py (hunk 2 above)
Comment (garrettwrong) [reply src/aspire/basis/fspca.py]: Sure, reworded..

FWIW, I think the `0` bit comes from cov2d historically crashing for clean images.  I think it has a special branch when identically zero that will skip some application of the noise model.

File: src/aspire/classification/rir_class2d.py (hunk 3 above)
Comment (garrettwrong) [reply src/aspire/classification/rir_class2d.py]: Thanks. That's how you know its handmade :D

File: src/aspire/classification/rir_class2d.py (hunk 4 above)
Comment (garrettwrong) [reply src/aspire/classification/rir_class2d.py]: thanks

File: tests/test_class2D.py (hunk 6 above)
Comment (garrettwrong) [reply tests/test_class2D.py]: Copy paste oversight, rm'd. Thanks

File: src/aspire/classification/rir_class2d.py (hunk 5 above)
Comment (garrettwrong) [reply src/aspire/classification/rir_class2d.py:147]: Mostly right, couple things.

FWIW, that's old code, and unchanged here, other than to move it after some other checks that should come first.  I appreciate you reviewing it anyway.

I think the only non trivial logic is differentiating whether the 400 comes from an arg default, or the user input exactly 400.  When it is an arg default I allow the variable to be inferred from other args.  I don't think we will get away from the combinatorial cases, but I can certainly use `None` and rewrite the logic. I'm not sure it will be the improvement you seek, but I will do it...


---

The 400 comes from the researcher (and I concur) in practice, that is outside of unit tests and degenerate inputs, it is a reasonable value for the RIRClass2D algo.  I don't think the same sort of thing holds for the FSPCABasis, which is ostensibly just pca applied to some other resolution based basis (FFB2D in this case).

Comment (garrettwrong) [reply src/aspire/classification/rir_class2d.py:147]: Let me know if this change was any better. Thanks!

File: src/aspire/basis/fspca.py [hunk 8]
```diff
@@ -38,11 +38,12 @@ def __init__(
 
         :param src: Source instance
         :param basis: Optional Fourier Bessel Basis (usually FFBBasis2D)
-        :param components: Number of components (compression).
-        Defaults None implies basis.count, therefor no compression.
-        :param noise_var: None estimates noise (default).
-        0 forces "clean" treatment (no weighting).
-        Other values assigned to noise_var.
+        :param components: Optionally assign number of principle components
```
Comment (chris-langfield) [inline src/aspire/basis/fspca.py]: principal* sorry to be pedantic 😁

File: src/aspire/classification/rir_class2d.py (hunk 5 above)
Comment (chris-langfield) [reply src/aspire/classification/rir_class2d.py:147]: Yeah I think this is good! Thanks

**Review by chris-langfield — DISMISSED** [review]
Just one typo but otherwise ready, thanks!

File: src/aspire/basis/fspca.py (hunk 8 above)
Comment (garrettwrong) [reply src/aspire/basis/fspca.py]: whoops 😊 , thanks

**Comment by garrettwrong** [comment]
Rebased with latest develop.

**Review by janden — APPROVED** [review]
Looks good!
